disnake/ext/commands/params.py

```diff
@@ -1115,10 +1253,22 @@ def converter_method(function: Any) -> ConverterMethod:
         return ConverterMethod(function)
 
 
-def register_injection(function: CallableT) -> CallableT:
+def register_injection(
+    function: Callable, *, autocompleters: Dict[str, Callable] = None
+) -> Injection:
```

Comment:
This currently returns an untyped Injection object, but that can be easily solved with the following.

```suggestion
def register_injection(
    function: Callable[P, T_], *, autocompleters: Dict[str, Callable] = None
) -> Injection[P, T_]:
```

<sub>but this reveals a typing issue with cogs and self, @shiftinv please help</sub>

Reply:
Yea, noticed that as well, I meant to fix that in another PR since it requires messing around with protocols.

Reply:
What about deprecating register_injection and recommending Injection.register over it? AFAIK the latter does not have this issue

Reply:
> Yea, noticed that as well, I meant to fix that in another PR since it requires messing around with protocols.

Do you have a fix for that yet?

Reply:
The issue is that the self type needs to be stripped from the paramspec, somewhat similar to
https://github.com/DisnakeDev/disnake/blob/74baf5023e6583722a34f386591a94a257244857/disnake/ext/commands/core.py#L1562-L1573
There's actually no need for protocols here though since we don't care about the self type, we just want to get rid of it, so this works: https://github.com/ItsAleph/disnake/compare/injected-autocomp...shiftinv:disnake:tmp/fix/injection-callback-typing
I can push it to this branch if that's fine with yall, otherwise it'll be a separate PR.

Reply:
go ahead, i think

Reply:
Alright, done. Those calls should now be typed properly, though I'd appreciate if someone else tests the typing as well.

Reply:
They are 👍